tests/components/qwikswitch/test_init.py

```diff
@@ -1,92 +1,81 @@
-
-    aoihttp mocker uses decode to fetch the next value.
-    """
-
-    def decode(self, _):
-        """Return next item from list."""
-        try:
-            res = list.pop(self, 0)
-            _LOGGER.debug("MockResponseList popped %s: %s", res, self)
-            if isinstance(res, Exception):
-                raise res
-            return res
-        except IndexError:
-            raise AssertionError("MockResponseList empty")
-
-    async def wait_till_empty(self, hass):
-        """Wait until empty."""
-        while self:
-            await hass.async_block_till_done()
-        await hass.async_block_till_done()
-
-
-LISTEN = AiohttpClientMockResponseList()
-
-
-@pytest.fixture
-def aioclient_mock():
-    """HTTP client listen and devices."""
-    devices = """[
-        {"id":"@000001","name":"Switch 1","type":"rel","val":"OFF",
-        "time":"1522777506","rssi":"51%"},
-        {"id":"@000002","name":"Light 2","type":"rel","val":"ON",
-        "time":"1522777507","rssi":"45%"},
-        {"id":"@000003","name":"Dim 3","type":"dim","val":"280c00",
-        "time":"1522777544","rssi":"62%"}]"""
-
-    with mock_aiohttp_client() as mock_session:
-        mock_session.get("http://127.0.0.1:2020/&listen", content=LISTEN)
-        mock_session.get("http://127.0.0.1:2020/&device", text=devices)
-        yield mock_session
-
-
-async def test_binary_sensor_device(hass, aioclient_mock):  # noqa: F811
+DEVICES = [
+    {
+        "id": "@000001",
+        "name": "Switch 1",
+        "type": "rel",
+        "val": "OFF",
+        "time": "1522777506",
+        "rssi": "51%",
+    },
+    {
+        "id": "@000002",
+        "name": "Light 2",
+        "type": "rel",
+        "val": "ON",
+        "time": "1522777507",
+        "rssi": "45%",
+    },
+    {
+        "id": "@000003",
+        "name": "Dim 3",
+        "type": "dim",
+        "val": "280c00",
+        "time": "1522777544",
+        "rssi": "62%",
+    },
+]
+
+
+async def test_binary_sensor_device(hass, aioclient_mock):
     """Test a binary sensor device."""
     config = {
         "qwikswitch": {
             "sensors": {"name": "s1", "id": "@a00001", "channel": 1, "type": "imod"}
         }
     }
+    aioclient_mock.get("http://127.0.0.1:2020/&device", json=DEVICES)
+    listen_mock = MockLongPollSideEffect()
+    aioclient_mock.get("http://127.0.0.1:2020/&listen", side_effect=listen_mock)
     await async_setup_component(hass, QWIKSWITCH, config)
+    hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
     await hass.async_block_till_done()
 
     state_obj = hass.states.get("binary_sensor.s1")
     assert state_obj.state == "off"
 
-    hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
-
-    LISTEN.append('{"id":"@a00001","cmd":"","data":"4e0e1601","rssi":"61%"}')
-    LISTEN.append(ClientError())  # Will cause a sleep
-
+    listen_mock.queue_response(
+        json={"id": "@a00001", "cmd": "", "data": "4e0e1601", "rssi": "61%"}
+    )
+    await asyncio.sleep(0.01)
```

Comment:
Is this still necessary if you block till done below?

Reply:
Nevermind, saw the other comment.

Reply:
i believe it is since these are tasks created by the library, so hass may have no pending task and return immediately if i understand it correctly

Reply:
BTW, we could add something to block_till_done that also monitors the external tasks on the event loop. However, it would need to know how many of these are running (for example, i noticed that qwikswitch always has a running task), so it would probably be useful only for tests. For tests it can be useful to see that there are no tasks that are not waiting on the event loop.
What do you think?

Reply:
we can't monitor the external tasks as that would also track the task that runs the current test that is awaiting all tasks :)

We only monitor any task that comes in via the `hass.async_create_task`, `hass.async_add_job` etc commands. `hass.loop.create_task` is not tracked as that's called directly on the loop.